### PR TITLE
Fix decode error when "NULL" value in the column with "primary key" flag

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/5859

Problem Summary:
TiFlash assumes all values in the column with "primary key" flag are non-null value.

However, users could create such "NULL" value in the "primary key" column by following steps

```sql
-- create table with `source` is nullable
create table test.t_case (
`case_no` varchar(32) not null,
`source` varchar(20) default null,
`p` varchar(12) DEFAULT NULL,
primary key (`case_no`)
);

-- insert some data and left `source` to be empty
insert into test.t_case(case_no)
values ("1"), ("2"), ("3"), ("4");

-- drop the primary key, fill the `source` to be non-empty
alter table t_case drop primary key;
update test.t_case set `source` = '' where `source` is NULL;
-- add new primary key with case_no and source
alter table t_case add primary key (case_no, source);

-- send the snapshot data to tiflash
alter table t_case set tiflash replica 1;
```

### What is changed and how it works?

* mainly change the logic in `addDefaultValueToColumnIfPossible`
  * For non-clustered index table(is_common_handle=false && pk_is_handle=false), fill default value even if the column contains primary key flag when the column is not appear in the encoded value(those rows are not supposed to be return to users)
  * For clustered index/pk_is_handle table, ignore the primary key column as they are supposed to be decoded from the encoded key
* Refine the exception message in `insertRangeFrom`
* Add more region info when exception thrown in `preHandleSnapshot`/`ingestSST`

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix the panic issue after creating the primary index with a column containing `NULL` value
```

### CODE REVIEW

File: tests/fullstack-test2/ddl/alter_pk.test [hunk 1]
```diff
@@ -57,3 +57,56 @@ mysql> alter table test.t drop primary key;
...
+## drop the primary key, fill the `source` to be non-empty
+## add new primary key with case_no and source
+mysql> alter table test.t_case drop primary key;
+mysql> update test.t_case set `source` = '' where `source` is NULL;
+mysql> alter table test.t_case add primary key (`case_no`, `source`);
+
+## send the snapshot data to tiflash
+mysql> alter table test.t_case set tiflash replica 1;
+func> wait_table test t_case
+mysql> select * from test.t_case;
```
Comment (JaySon-Huang) [inline tests/fullstack-test2/ddl/alter_pk.test:77]:
```suggestion
mysql> select case_no,p,source from test.t_case;
```

File: tests/fullstack-test2/ddl/alter_pk.test [hunk 2]
```diff
@@ -57,3 +57,56 @@ mysql> alter table test.t drop primary key;
...
+
+mysql> alter table test.t_case add column `source` varchar(20) not null;
+## drop the primary key, add new primary key with case_no and source
+mysql> alter table test.t_case drop primary key;
+mysql> alter table test.t_case add primary key (`case_no`, `source`);
+
+## send the snapshot data to tiflash
+mysql> alter table test.t_case set tiflash replica 1;
+func> wait_table test t_case
+mysql> select * from test.t_case;
```
Comment (JaySon-Huang) [inline tests/fullstack-test2/ddl/alter_pk.test:102]:
```suggestion
mysql> select case_no,p,source from test.t_case;
```

**Comment by JaySon-Huang** [comment]
The helper functions in `RegionBlockReaderTestFixture` is buggy when testing with primary key columns. I will split it into another PR to refine the test suite first.

File: dbms/src/Storages/Transaction/RowCodec.cpp [hunk 3]
```diff
@@ -285,53 +290,96 @@ void encodeRowV2(const TiDB::TableInfo & table_info, const std::vector<Field> &
...
+        return is_big ? appendRowV2ToBlockImpl<true>(raw_value, column_ids_iter, column_ids_iter_end, block, block_column_pos, column_infos, pk_handle_id, fill_missing_column_with_pk_flag, force_decode)
+                      : appendRowV2ToBlockImpl<false>(raw_value, column_ids_iter, column_ids_iter_end, block, block_column_pos, column_infos, pk_handle_id, fill_missing_column_with_pk_flag, force_decode);
+    }
+    default:
+        return appendRowV1ToBlock(raw_value, column_ids_iter, column_ids_iter_end, block, block_column_pos, column_infos, pk_handle_id, fill_missing_column_with_pk_flag, force_decode);
+    }
 }
 
-inline bool addDefaultValueToColumnIfPossible(const ColumnInfo & column_info, Block & block, size_t block_column_pos, bool force_decode)
+inline bool addDefaultValueToColumnIfPossible(const ColumnInfo & column_info, Block & block, size_t block_column_pos, bool fill_column_with_pk_flag, bool force_decode)
```
Comment (hongyunyan) [inline dbms/src/Storages/Transaction/RowCodec.cpp:356]: I think the variable name "fill_column_with_pk_flag" may by a little difficult to understand.

Comment (JaySon-Huang) [reply dbms/src/Storages/Transaction/RowCodec.cpp:356]: what about `table_is_non_clustered`? The name means that the keys for row data consist of internal `_tidb_rowid` implicitly assigned by TiDB, and we should fill the missing value for those columns with primary key flags under this situation

https://docs.pingcap.com/tidb/dev/clustered-indexes#clustered-indexes

Comment (lidezhu) [reply dbms/src/Storages/Transaction/RowCodec.cpp:356]: How about `ignore_pk_if_absent`?

Comment (hongyunyan) [reply dbms/src/Storages/Transaction/RowCodec.cpp:356]: I think `ignore_pk_if_absent` is more straight forward to understand.

**Review by lidezhu — APPROVED** [review]
LGTM

File: dbms/src/Storages/Transaction/RowCodec.cpp (hunk 3 above)
Comment (JaySon-Huang) [reply dbms/src/Storages/Transaction/RowCodec.cpp]: updated

**Review by hongyunyan — APPROVED** [review]
LGTM

File: dbms/src/Storages/Transaction/RowCodec.cpp [hunk 4]
```diff
@@ -290,53 +290,96 @@ void encodeRowV2(const TiDB::TableInfo & table_info, const std::vector<Field> &
...
+        if (ignore_pk_if_absent)
+            return true;
+
+        assert(!ignore_pk_if_absent);
+        if (!force_decode)
+            return false;
+        // Else non-clustered index, and not pk_is_handle, it could be a row encoded by older schema,
+        // we need to fill the column wich has primary key flag with default value.
+        // fallthrough to fill default value when force_decode
+        LOG_FMT_WARNING(Logger::get("RowCodec"), "fill default value for missing primary key column, column_id={} column_name={}", column_info.id, column_info.name);
```
Comment (flowbehappy) [inline dbms/src/Storages/Transaction/RowCodec.cpp:373]: I'm afraid this warning too chatty....

Comment (flowbehappy) [reply dbms/src/Storages/Transaction/RowCodec.cpp:373]: I suggest using trace

**Review by flowbehappy — APPROVED** [review]
LGTM with a comment

File: dbms/src/Storages/Transaction/RowCodec.cpp (hunk 4 above)
Comment (JaySon-Huang) [inline dbms/src/Storages/Transaction/RowCodec.cpp:373]:
```suggestion
```

Comment (JaySon-Huang) [reply dbms/src/Storages/Transaction/RowCodec.cpp]: removed

**Comment by breezewish** [comment]
This PR's description really refreshed my mind LOL

**Pull request merged**